lib/services/access_request.go

```diff
@@ -2032,6 +2252,80 @@ func countAllowedLogins(role types.Role) int {
 	return len(allowed)
 }
 
+// getKubeResourceKinds just extracts the kinds from the list.
+// If a wildcard is present, then all supported resource types are returned.
+func getKubeResourceKinds(requestModes []types.RequestModeKubernetesResource) []string {
+	var kinds []string
+	for _, rm := range requestModes {
+		if rm.Kind == types.Wildcard {
+			return types.KubernetesResourcesKinds
+		}
+		kinds = append(kinds, rm.Kind)
+	}
+	return kinds
+}
+
+// getAllowedKubeResourceKinds returns only the allowed kinds that were not in the
+// denied list.
+func getAllowedKubeResourceKinds(allowedKinds []string, deniedKinds []string) []string {
+	denied := make(map[string]struct{})
+	for _, kind := range deniedKinds {
+		denied[kind] = struct{}{}
+	}
+
+	allowed := make([]string, 0, len(allowedKinds))
+	for _, kind := range allowedKinds {
+		if _, denied := denied[kind]; !denied {
+			allowed = append(allowed, kind)
+		}
+	}
+	return apiutils.Deduplicate(allowed)
+}
+
+// matchKubeResourcesWithRequestModes goes through each resource IDs and finds matching request modes marked in the "seenRequestedKinds" map.
+// It returns true if there was at least one match found.
+// It also returns all the allowable kinds that the user can request, subtracting any denied kinds from it.
+func (m *RequestValidator) matchKubeResourcesWithRequestModes(roleName string, kubernetesResources []types.KubernetesResource, seenRequestedKinds map[string]bool) ([]string, bool) {
```

Comment:
I'm sorry but this comment does not really help me understand the function. Can you explain what seenResourceKinds is, and it returns true if there is a least one "match" which is what?

Reply:
tried to clarify it further